
"Ce projet vise à analyser les données d’opportunités de vente issues d’un système de gestion de la relation client (CRM). L’ensemble des données, provenant de Maven Analytics, contient des informations détaillées sur les opportunités de vente, le potentiel de revenus et les interactions avec les clients, la performance globale des ventes ."

In [1]:
pip install pandas 

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
pip install numpy


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import pandas as pd

import numpy as np

In [4]:
accounts_data= pd.read_csv("accounts.csv")

In [5]:
accounts_data.head(50)

account              sector  year_established  \
0               Acme Corporation           technolgy              1996   
1                     Betasoloin             medical              1999   
2                       Betatech             medical              1986   
3                     Bioholding             medical              2012   
4                        Bioplex             medical              1991   
5                       Blackzim              retail              2009   
6                  Bluth Company           technolgy              1993   
7                     Bubba Gump            software              2002   
8                        Cancity              retail              2001   
9                         Cheers       entertainment              1993   
10                       Codehow            software              1998   
11                        Condax             medical              2017   
12                       Conecom           technolgy              2005   
13                Dalttechnology            software              2013   
14                       dambase           marketing              1995   
15                       Domzoom       entertainment              1998   
16                        Doncon           technolgy              2010   
17                   Donquadtech           technolgy              1992   
18                      Dontechi            software              1982   
19                       Donware           marketing              1999   
20                    Fasehatice              retail              1990   
21                      Faxquote  telecommunications              1995   
22                       Finhigh             finance              2006   
23                        Finjob          employment              1988   
24                    Funholding             finance              1991   
25                     Ganjaflex              retail              1995   
26                    Gekko & Co              retail              1990   
27  Genco Pura Olive Oil Company              retail              2007   
28            Globex Corporation           technolgy              2000   
29                      Gogozoom  telecommunications              2007   
30                       Golddex             finance              2008   
31                    Goodsilron           marketing              2000   
32                    Green-Plus            services              2003   
33                  Groovestreet              retail              2003   
34                        Hatfan            services              1982   
35                      Hottechi           technolgy              1997   
36                       Initech  telecommunications              1994   
37                         Inity           marketing              1986   
38                         Isdom             medical              2002   
39                   Iselectrics           technolgy              2011   
40                       J-Texon              retail              1989   
41                      Kan-code            software              1982   
42                    Kinnamplus              retail              2004   
43                         Konex           technolgy              1980   
44                     Konmatfix           marketing              1985   
45                      Labdrill             medical              1985   
46                    Lexiqvolax             medical              2004   
47               Massive Dynamic       entertainment              1989   
48                     Mathtouch           marketing              1984   
49                       Nam-zim            services              1987   

     revenue  employees office_location     subsidiary_of  
0    1100.04       2822   United States               NaN  
1     251.41        495   United States               NaN  
2     647.18       1185           Kenya               NaN  
3     587.

In [6]:
lignes=accounts_data.shape[0]
variables=accounts_data.shape[1]

print(f"le nombre de lignes dans la base accounts_data est {lignes} et le nombre de variable est {variables}")

le nombre de lignes dans la base accounts_data est 85 et le nombre de variable est 7


Exploaration accounts_data

In [7]:
dictionnaire ={
    'account':"compte de l'entreprise ",
    'sector':"secteurs d'activité de l'entreprise",
    'year_established':"année de creation de l'entreprise",
    'revenue':"montant total des revenu générés par l'entreprise",
    'employees':"le nombre total d'employés travaillant dans l'entreprise",
    'office_location':"la localisation géographique principale du bureau de l'entreprise",
    'subsidiary_of':"filiale d'une autre société"
}


def info(data):
    Information=pd.DataFrame({
        'variables':data.columns,
        'types':data.dtypes,
        'unique_value':data.nunique(),
        'Na_counts':data.isna().sum(),
        'Na_percents%':data.isna().mean().round(4)*100

        
    }).reset_index(drop=True)

    Information['variables_descriptives']= Information['variables'].map( dictionnaire)

    return Information

metadonnées1=info(accounts_data)
metadonnées1

variables    types  unique_value  Na_counts  Na_percents%  \
0           account   object            85          0          0.00   
1            sector   object            10          0          0.00   
2  year_established    int64            35          0          0.00   
3           revenue  float64            85          0          0.00   
4         employees    int64            85          0          0.00   
5   office_location   object            15          0          0.00   
6     subsidiary_of   object             7         70         82.35   

                              variables_descriptives  
0                            compte de l'entreprise   
1                secteurs d'activité de l'entreprise  
2                  année de creation de l'entreprise  
3  montant total des revenu générés par l'entreprise  
4  le nombre total d'employés travaillant dans l'...  
5  la localisation géographique principale du bur...  
6                        filiale d'une autre société

In [8]:
accounts_data["subsidiary_of"]=accounts_data["subsidiary_of"].fillna("Indépendant")


In [9]:
metadonnées1

variables    types  unique_value  Na_counts  Na_percents%  \
0           account   object            85          0          0.00   
1            sector   object            10          0          0.00   
2  year_established    int64            35          0          0.00   
3           revenue  float64            85          0          0.00   
4         employees    int64            85          0          0.00   
5   office_location   object            15          0          0.00   
6     subsidiary_of   object             7         70         82.35   

                              variables_descriptives  
0                            compte de l'entreprise   
1                secteurs d'activité de l'entreprise  
2                  année de creation de l'entreprise  
3  montant total des revenu générés par l'entreprise  
4  le nombre total d'employés travaillant dans l'...  
5  la localisation géographique principale du bur...  
6                        filiale d'une autre société

In [10]:
# affiche tout   les valeurs unique dans la colonne

accounts_data["subsidiary_of"].unique()  

array(['Indépendant', 'Acme Corporation', 'Massive Dynamic', 'Bubba Gump',
       'Inity', 'Sonron', 'Golddex', 'Warephase'], dtype=object)

In [11]:
# pour compter le nombre de valeurs manquantes dans la colonne 

print(accounts_data["subsidiary_of"].isnull().sum())  


0


In [12]:
# nombre de filiales et des independants , en comptant le nombre de frequence de chaque valeur unique dans la colonne 

filiales_counts=accounts_data["subsidiary_of"].value_counts()

print((filiales_counts))

subsidiary_of
Indépendant         70
Acme Corporation     4
Sonron               3
Bubba Gump           2
Inity                2
Golddex              2
Massive Dynamic      1
Warephase            1
Name: count, dtype: int64


In [13]:
# visualisation avec ploty
import plotly.express as px
import pandas as pd


# Préparer un DataFrame pour Plotly
filiales_counts_df = filiales_counts.reset_index()
filiales_counts_df.columns = ["Type d'entreprise", "Nombre d'entreprises"]  

# Créer un graphique à barres avec Plotly
fig = px.bar(
    filiales_counts_df,
    x="Type d'entreprise",  
    y="Nombre d'entreprises",  
    title="Répartition des Indépendants et Filiales",
    labels={"Type d'entreprise", "Nombre d'entreprises"},
    text="Nombre d'entreprises"  
)

# Améliorer la présentation
fig.update_traces(textposition="outside")  
fig.update_layout(xaxis=dict(tickangle=45))  

# Afficher le graphique
fig.show()


In [14]:
#analyse des revenues de filiales et independants

independants_revenus=accounts_data[accounts_data["subsidiary_of"]=="Indépendant"]['revenue'].mean()
filiales_revenus=accounts_data[accounts_data["subsidiary_of"]!= "Indépendant"]['revenue'].mean()

print(f'le revenu  moyen des independants : {independants_revenus: .2f}')

print(f'le revenu  moyen des filiales  :{filiales_revenus: .2f}')

le revenu  moyen des independants :  1990.77
le revenu  moyen des filiales  : 2012.65


In [15]:
#visualisation avec ploty

import plotly.express as px
import pandas as pd

# Préparer les données pour la visualisation
data_visualisation = pd.DataFrame({
    "Type d'entreprise": ["Indépendants", "Filiales"],
    "Revenu moyen": [independants_revenus, filiales_revenus]
})

# Créer un graphique à barres
fig = px.bar(
    data_visualisation,
    x="Type d'entreprise",
    y="Revenu moyen",
    title="Revenu moyen : Indépendants vs Filiales",
    text="Revenu moyen",  
    labels={"Revenu moyen": "Revenu moyen ($)", "Type d'entreprise": "Type d'entreprise"}
)

# Personnalisation esthétique
fig.update_traces(
    marker=dict(color=['lightblue', 'blue']), 
    texttemplate='%{text:.2f}',               
    textposition='outside'                    
)
fig.update_layout(
    title_font_size=20,                       
    xaxis=dict(title_font_size=16),           
    yaxis=dict(title_font_size=16)            
)

# Afficher le graphique
fig.show()


In [16]:
# verifions si notre data ne contient pas de données manquantes
accounts_data.isnull().sum()



account             0
sector              0
year_established    0
revenue             0
employees           0
office_location     0
subsidiary_of       0
dtype: int64

exploration Métadonnées sales_pipeline 

In [17]:
Sales_pipeline=pd.read_csv("sales_pipeline.csv")

In [18]:
Nombre_lignes=Sales_pipeline.shape[0]
Nombre_variables=Sales_pipeline.shape[1]

print(f"le nombre de ligne dans la base de données : {Nombre_lignes}")

print(f"le nombre de variables dans la base de données : {Nombre_variables}")

le nombre de ligne dans la base de données : 8800
le nombre de variables dans la base de données : 8


In [19]:
dictionnaire ={
    "opportunity_id": "ID de l'opportunité",
    "sales_agent": "Agent commercial",
    "product": "Produit",
    "account": "Compte",
    "deal_stage": "Étape de l'affaire",
    "engage_date": "Date d'engagement",
    "close_date": "Date de clôture",
    "close_value": "Valeur de clôture"
}


def information(data):
    Information=pd.DataFrame({
        'variables':data.columns,
        'types':data.dtypes,
        'unique_value':data.nunique(),
        'Na_counts':data.isna().sum(),
        'Na_percents%':data.isna().mean().round(4)*100

        
    }).reset_index(drop=True)

    Information['variables_descriptives']= Information['variables'].map( dictionnaire)

    return Information

metadonnées2=info(Sales_pipeline)
metadonnées2

variables    types  unique_value  Na_counts  Na_percents%  \
0  opportunity_id   object          8800          0          0.00   
1     sales_agent   object            30          0          0.00   
2         product   object             7          0          0.00   
3         account   object            85       1425         16.19   
4      deal_stage   object             4          0          0.00   
5     engage_date   object           421        500          5.68   
6      close_date   object           306       2089         23.74   
7     close_value  float64          2051       2089         23.74   

  variables_descriptives  
0    ID de l'opportunité  
1       Agent commercial  
2                Produit  
3                 Compte  
4     Étape de l'affaire  
5      Date d'engagement  
6        Date de clôture  
7      Valeur de clôture

In [20]:
print(Sales_pipeline.isnull().sum())


opportunity_id       0
sales_agent          0
product              0
account           1425
deal_stage           0
engage_date        500
close_date        2089
close_value       2089
dtype: int64


In [21]:
print(Sales_pipeline.isnull().sum())

opportunity_id       0
sales_agent          0
product              0
account           1425
deal_stage           0
engage_date        500
close_date        2089
close_value       2089
dtype: int64


In [22]:
metadonnées2

variables    types  unique_value  Na_counts  Na_percents%  \
0  opportunity_id   object          8800          0          0.00   
1     sales_agent   object            30          0          0.00   
2         product   object             7          0          0.00   
3         account   object            85       1425         16.19   
4      deal_stage   object             4          0          0.00   
5     engage_date   object           421        500          5.68   
6      close_date   object           306       2089         23.74   
7     close_value  float64          2051       2089         23.74   

  variables_descriptives  
0    ID de l'opportunité  
1       Agent commercial  
2                Produit  
3                 Compte  
4     Étape de l'affaire  
5      Date d'engagement  
6        Date de clôture  
7      Valeur de clôture

In [23]:
# Ensure 'engage_date' and 'close_date' are properly converted to datetime
Sales_pipeline["engage_date"] = pd.to_datetime(Sales_pipeline["engage_date"], errors="coerce").dt.normalize()
Sales_pipeline["close_date"] = pd.to_datetime(Sales_pipeline["close_date"], errors="coerce").dt.normalize()

# Check and fill missing values in the sales pipeline
def fill_missing_sales_pipeline(df):
    df["account"].fillna("Inconnu", inplace=True)
    df["close_value"].fillna(df["close_value"].median(), inplace=True)
    
    # Imputation conditionnelle pour les dates en fonction de 'sales_agent'
    mean_engage_date = df.groupby("sales_agent")["engage_date"].transform("mean")
    df["engage_date"].fillna(mean_engage_date, inplace=True)

    mean_close_date = df.groupby("sales_agent")["close_date"].transform("mean")
    df["close_date"].fillna(mean_close_date, inplace=True)

    return df

sales_pipeline = fill_missing_sales_pipeline(Sales_pipeline)

C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\1965180521.py:7: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\1965180521.py:8: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=

In [24]:
Sales_pipeline.dtypes

opportunity_id            object
sales_agent               object
product                   object
account                   object
deal_stage                object
engage_date       datetime64[ns]
close_date        datetime64[ns]
close_value              float64
dtype: object

In [25]:
print(Sales_pipeline.isnull().sum())

opportunity_id    0
sales_agent       0
product           0
account           0
deal_stage        0
engage_date       0
close_date        0
close_value       0
dtype: int64


Fusion des données account_data pour enrichir le pipeline des ventes 

In [26]:
merged_data = Sales_pipeline.merge(accounts_data, on="account", how="left")


In [27]:
# Fill missing values in the merged dataframe
merged_data["sector"].fillna("Inconnu", inplace=True)
merged_data["revenue"].fillna(merged_data["revenue"].median(), inplace=True)
merged_data["year_established"].fillna(merged_data["year_established"].median(), inplace=True)
merged_data["employees"].fillna(merged_data["employees"].median(), inplace=True)
merged_data["office_location"].fillna("Non spécifié", inplace=True)
merged_data["subsidiary_of"].fillna("Indépendant", inplace=True)


C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\2518527563.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\2518527563.py:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=

In [28]:
# Convertir en entier après remplacement
merged_data["year_established"] = merged_data["year_established"].astype(int)
merged_data["employees"] = merged_data["employees"].astype(int)

In [29]:
# Conversion en datetime
merged_data["engage_date"] = pd.to_datetime(merged_data["engage_date"], errors="coerce")
merged_data["close_date"] = pd.to_datetime(merged_data["close_date"], errors="coerce")

# Vérification des types avant formatage
print(merged_data["engage_date"].dtypes)  




datetime64[ns]


In [30]:
print(merged_data["close_date"].dtypes) 

datetime64[ns]


In [31]:
print(merged_data["engage_date"].dtypes) 

datetime64[ns]


In [32]:
# Final check for missing values
print(merged_data.isnull().sum())

opportunity_id      0
sales_agent         0
product             0
account             0
deal_stage          0
engage_date         0
close_date          0
close_value         0
sector              0
year_established    0
revenue             0
employees           0
office_location     0
subsidiary_of       0
dtype: int64


In [33]:
merged_data

opportunity_id        sales_agent         product  account   deal_stage  \
0          1C1I7A6R        Moses Frase  GTX Plus Basic  Cancity          Won   
1          Z063OYW0    Darcel Schlecht          GTXPro    Isdom          Won   
2          EC4QE1BX    Darcel Schlecht      MG Special  Cancity          Won   
3          MV1LWRNH        Moses Frase       GTX Basic  Codehow          Won   
4          PE84CX4O          Zane Levy       GTX Basic   Hatfan          Won   
...             ...                ...             ...      ...          ...   
8795       9MIWFW5J  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8796       6SLKZ8FI  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8797       LIB4KUZJ  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8798       18IUIUK0  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8799       8I5ONXJX  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   

                       engage_date                    close_date  close_value  \
0    2016-10-20 00:00:00.000000000 2017-03-01 00:00:00.000000000       1054.0   
1    2016-10-25 00:00:00.000000000 2017-03-11 00:00:00.000000000       4514.0   
2    2016-10-25 00:00:00.000000000 2017-03-07 00:00:00.000000000         50.0   
3    2016-10-25 00:00:00.000000000 2017-03-09 00:00:00.000000000        588.0   
4    2016-10-25 00:00:00.000000000 2017-03-02 00:00:00.000000000        517.0   
...                            ...                           ...          ...   
8795 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8796 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8797 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8798 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8799 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   

        sector  year_established  revenue  employees office_location  \
0       retail              2001   718.62       2448   United States   
1      medical              2002  3178.24       4540   United States   
2       retail              2001   718.62       2448   United States   
3     software              1998  2714.90       2641   United States   
4     services              1982   792.46       1299   United States   
...        ...               ...      ...        ...             ...   
8795   Inconnu              1995  1698.20       3492    Non spécifié   
8796   Inconnu              1995  1698.20       3492    Non spécifié   
8797   Inconnu              1995  1698.20       3492    Non spécifié   
8798   Inconnu              1995  1698.20       3492    Non spécifié   
8799   Inconnu              1995  1698.20       3492    Non spécifié   

         subsidiary_of  
0          Indépendant  
1          Indépendant  
2          Indépendant  
3     Acme Corporation  
4          Indépendant  
...                ...  
8795       Indépendant  
8796       Indépendant  
8797       Indépendant  
8798       Indépendant  
8799       Indépendant  

[8800 rows x 14 columns]

In [34]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8800 entries, 0 to 8799
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   opportunity_id    8800 non-null   object        
 1   sales_agent       8800 non-null   object        
 2   product           8800 non-null   object        
 3   account           8800 non-null   object        
 4   deal_stage        8800 non-null   object        
 5   engage_date       8800 non-null   datetime64[ns]
 6   close_date        8800 non-null   datetime64[ns]
 7   close_value       8800 non-null   float64       
 8   sector            8800 non-null   object        
 9   year_established  8800 non-null   int32         
 10  revenue           8800 non-null   float64       
 11  employees         8800 non-null   int32         
 12  office_location   8800 non-null   object        
 13  subsidiary_of     8800 non-null   object        
dtypes: datetime64[ns](2), fl

In [35]:
# verifions s'il ya des doublons

merged_data.duplicated().sum()



0

Remplaçons les valeurs manquantes dans les colonnes de accounts_data par valeurs par défaut ou calculées

In [36]:
merged_data["sector"].fillna("Inconnu", inplace=True)
merged_data["revenue"].fillna(merged_data["revenue"].median(), inplace=True)
merged_data["year_established"].fillna(merged_data["year_established"].median(), inplace=True)
merged_data["employees"].fillna(merged_data["employees"].median(), inplace=True)
merged_data["office_location"].fillna("Non spécifié", inplace=True)
merged_data["subsidiary_of"].fillna("Indépendant", inplace=True)


C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\3525728804.py:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\3525728804.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=

In [37]:
print(merged_data.isnull().sum())

opportunity_id      0
sales_agent         0
product             0
account             0
deal_stage          0
engage_date         0
close_date          0
close_value         0
sector              0
year_established    0
revenue             0
employees           0
office_location     0
subsidiary_of       0
dtype: int64


In [38]:
print(merged_data["engage_date"].dtype)

datetime64[ns]


In [39]:
merged_data.dtypes

opportunity_id              object
sales_agent                 object
product                     object
account                     object
deal_stage                  object
engage_date         datetime64[ns]
close_date          datetime64[ns]
close_value                float64
sector                      object
year_established             int32
revenue                    float64
employees                    int32
office_location             object
subsidiary_of               object
dtype: object

Analyse temporelle

In [40]:
# Extraction des composantes temporelles
merged_data["engage_day"] = merged_data["engage_date"].dt.strftime("%A") 
merged_data["engage_month"] = merged_data["engage_date"].dt.strftime("%B")
merged_data["engage_year"] = merged_data["engage_date"].dt.year  
merged_data["engage_quarter"] = merged_data["engage_date"].dt.quarter  

merged_data["close_day"] = merged_data["close_date"].dt.strftime("%A")
merged_data["close_month"] = merged_data["close_date"].dt.strftime("%B") 
merged_data["close_year"] = merged_data["close_date"].dt.year
merged_data["close_quarter"] = merged_data["close_date"].dt.quarter

# Vérification des colonnes nouvellement ajoutées
print(merged_data[["engage_year", "engage_month", "engage_day", "engage_quarter"]].head())
print(merged_data[["close_year", "close_month", "close_day", "close_quarter"]].head())


   engage_year engage_month engage_day  engage_quarter
0         2016      October   Thursday               4
1         2016      October    Tuesday               4
2         2016      October    Tuesday               4
3         2016      October    Tuesday               4
4         2016      October    Tuesday               4
   close_year close_month  close_day  close_quarter
0        2017       March  Wednesday              1
1        2017       March   Saturday              1
2        2017       March    Tuesday              1
3        2017       March   Thursday              1
4        2017       March   Thursday              1


In [41]:
merged_data

opportunity_id        sales_agent         product  account   deal_stage  \
0          1C1I7A6R        Moses Frase  GTX Plus Basic  Cancity          Won   
1          Z063OYW0    Darcel Schlecht          GTXPro    Isdom          Won   
2          EC4QE1BX    Darcel Schlecht      MG Special  Cancity          Won   
3          MV1LWRNH        Moses Frase       GTX Basic  Codehow          Won   
4          PE84CX4O          Zane Levy       GTX Basic   Hatfan          Won   
...             ...                ...             ...      ...          ...   
8795       9MIWFW5J  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8796       6SLKZ8FI  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8797       LIB4KUZJ  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8798       18IUIUK0  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8799       8I5ONXJX  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   

                       engage_date                    close_date  close_value  \
0    2016-10-20 00:00:00.000000000 2017-03-01 00:00:00.000000000       1054.0   
1    2016-10-25 00:00:00.000000000 2017-03-11 00:00:00.000000000       4514.0   
2    2016-10-25 00:00:00.000000000 2017-03-07 00:00:00.000000000         50.0   
3    2016-10-25 00:00:00.000000000 2017-03-09 00:00:00.000000000        588.0   
4    2016-10-25 00:00:00.000000000 2017-03-02 00:00:00.000000000        517.0   
...                            ...                           ...          ...   
8795 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8796 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8797 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8798 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8799 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   

        sector  year_established  ...  office_location     subsidiary_of  \
0       retail              2001  ...    United States       Indépendant   
1      medical              2002  ...    United States       Indépendant   
2       retail              2001  ...    United States       Indépendant   
3     software              1998  ...    United States  Acme Corporation   
4     services              1982  ...    United States       Indépendant   
...        ...               ...  ...              ...               ...   
8795   Inconnu              1995  ...     Non spécifié       Indépendant   
8796   Inconnu              1995  ...     Non spécifié       Indépendant   
8797   Inconnu              1995  ...     Non spécifié       Indépendant   
8798   Inconnu              1995  ...     Non spécifié       Indépendant   
8799   Inconnu              1995  ...     Non spécifié       Indépendant   

     engage_day engage_month engage_year engage_quarter  close_day  \
0      Thursday      October        2016              4  Wednesday   
1       Tuesday      October        2016              4   Saturday   
2       Tuesday      October        2016              4    Tuesday   
3       Tuesday      October        2016              4   Thursday   
4       Tuesday      October        2016              4   Thursday   
...         ...          ...         ...            ...        ...   
8795     Friday         June        2017              2  Wednesday   
8796     Friday         June        2017              2  Wednesday   
8797     Friday         June        2017              2  Wednesday   
8798     Friday         June        2017              2  Wednesday   
8799     Friday         June        2017              2  Wednesday   

      close_month close_year close_quarter  
0           March       2017             1  
1           March       2017             1  
2           March       2017             1  
3           March       2017             1  
4           March       2017             1  
...           ...        ...           ...  
87

In [42]:
# Charger les fichiers CSV nécessaires
sales_teams = pd.read_csv('sales_teams.csv')
products = pd.read_csv('products.csv')

# Joindre les informations de 'sales_teams' à 'merged_data'
merged_data = merged_data.merge(
    sales_teams,
    how='left',  # Utilisez 'left' pour ne pas perdre de données de merged_data
    on='sales_agent'
)

# Joindre les informations de 'products' à 'merged_data'
merged_data = merged_data.merge(
    products,
    how='left',  # Utilisez 'left' pour ne pas perdre de données de merged_data
    on='product'
)

# Vérification des colonnes ajoutées et des valeurs manquantes après fusion
print(merged_data.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8800 entries, 0 to 8799
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   opportunity_id    8800 non-null   object        
 1   sales_agent       8800 non-null   object        
 2   product           8800 non-null   object        
 3   account           8800 non-null   object        
 4   deal_stage        8800 non-null   object        
 5   engage_date       8800 non-null   datetime64[ns]
 6   close_date        8800 non-null   datetime64[ns]
 7   close_value       8800 non-null   float64       
 8   sector            8800 non-null   object        
 9   year_established  8800 non-null   int32         
 10  revenue           8800 non-null   float64       
 11  employees         8800 non-null   int32         
 12  office_location   8800 non-null   object        
 13  subsidiary_of     8800 non-null   object        
 14  engage_day        8800 n

In [43]:
# Remplir les colonnes manquantes avec des valeurs par défaut
merged_data['series'].fillna("Inconnu", inplace=True)
merged_data['sales_price'].fillna(merged_data['sales_price'].median(), inplace=True)


C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\2125799713.py:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\2125799713.py:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=

In [44]:
# Convertir les colonnes en 'category'
merged_data['sector'] = merged_data['sector'].astype('category')
merged_data['deal_stage'] = merged_data['deal_stage'].astype('category')
merged_data['regional_office'] = merged_data['regional_office'].astype('category')

# Vérifiez les types après conversion
print(merged_data.dtypes)


opportunity_id              object
sales_agent                 object
product                     object
account                     object
deal_stage                category
engage_date         datetime64[ns]
close_date          datetime64[ns]
close_value                float64
sector                    category
year_established             int32
revenue                    float64
employees                    int32
office_location             object
subsidiary_of               object
engage_day                  object
engage_month                object
engage_year                  int32
engage_quarter               int32
close_day                   object
close_month                 object
close_year                   int32
close_quarter                int32
manager                     object
regional_office           category
series                      object
sales_price                float64
dtype: object


In [45]:
print(merged_data.isnull().sum())

opportunity_id      0
sales_agent         0
product             0
account             0
deal_stage          0
engage_date         0
close_date          0
close_value         0
sector              0
year_established    0
revenue             0
employees           0
office_location     0
subsidiary_of       0
engage_day          0
engage_month        0
engage_year         0
engage_quarter      0
close_day           0
close_month         0
close_year          0
close_quarter       0
manager             0
regional_office     0
series              0
sales_price         0
dtype: int64


In [46]:
merged_data

opportunity_id        sales_agent         product  account   deal_stage  \
0          1C1I7A6R        Moses Frase  GTX Plus Basic  Cancity          Won   
1          Z063OYW0    Darcel Schlecht          GTXPro    Isdom          Won   
2          EC4QE1BX    Darcel Schlecht      MG Special  Cancity          Won   
3          MV1LWRNH        Moses Frase       GTX Basic  Codehow          Won   
4          PE84CX4O          Zane Levy       GTX Basic   Hatfan          Won   
...             ...                ...             ...      ...          ...   
8795       9MIWFW5J  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8796       6SLKZ8FI  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8797       LIB4KUZJ  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8798       18IUIUK0  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   
8799       8I5ONXJX  Versie Hillebrand     MG Advanced  Inconnu  Prospecting   

                       engage_date                    close_date  close_value  \
0    2016-10-20 00:00:00.000000000 2017-03-01 00:00:00.000000000       1054.0   
1    2016-10-25 00:00:00.000000000 2017-03-11 00:00:00.000000000       4514.0   
2    2016-10-25 00:00:00.000000000 2017-03-07 00:00:00.000000000         50.0   
3    2016-10-25 00:00:00.000000000 2017-03-09 00:00:00.000000000        588.0   
4    2016-10-25 00:00:00.000000000 2017-03-02 00:00:00.000000000        517.0   
...                            ...                           ...          ...   
8795 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8796 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8797 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8798 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   
8799 2017-06-16 07:58:26.188925184 2017-07-26 03:54:32.727272704        472.0   

        sector  year_established  ...  engage_year  engage_quarter  close_day  \
0       retail              2001  ...         2016               4  Wednesday   
1      medical              2002  ...         2016               4   Saturday   
2       retail              2001  ...         2016               4    Tuesday   
3     software              1998  ...         2016               4   Thursday   
4     services              1982  ...         2016               4   Thursday   
...        ...               ...  ...          ...             ...        ...   
8795   Inconnu              1995  ...         2017               2  Wednesday   
8796   Inconnu              1995  ...         2017               2  Wednesday   
8797   Inconnu              1995  ...         2017               2  Wednesday   
8798   Inconnu              1995  ...         2017               2  Wednesday   
8799   Inconnu              1995  ...         2017               2  Wednesday   

     close_month close_year close_quarter           manager  regional_office  \
0          March       2017             1  Dustin Brinkmann          Central   
1          March       2017             1     Melvin Marxen          Central   
2          March       2017             1     Melvin Marxen          Central   
3          March       2017             1  Dustin Brinkmann          Central   
4          March       2017             1     Summer Sewald             West   
...          ...        ...           ...               ...              ...   
8795        July       2017             3  Dustin Brinkmann          Central   
8796        July       2017             3  Dustin Brinkmann          Central   
8797        July       2017             3  Dustin Brinkmann          Central   
8798        July       2017             3  Dustin Brinkmann          Central   
8799        July       2017             3  Dustin Brinkmann          Central   

       series sales_price  
0         GTX      1096.0  
1     Inconnu      1096.0  
2          MG        55.0  
3         GTX       550.0

In [47]:
merged_data.dtypes

opportunity_id              object
sales_agent                 object
product                     object
account                     object
deal_stage                category
engage_date         datetime64[ns]
close_date          datetime64[ns]
close_value                float64
sector                    category
year_established             int32
revenue                    float64
employees                    int32
office_location             object
subsidiary_of               object
engage_day                  object
engage_month                object
engage_year                  int32
engage_quarter               int32
close_day                   object
close_month                 object
close_year                   int32
close_quarter                int32
manager                     object
regional_office           category
series                      object
sales_price                float64
dtype: object

Chargement du fichier netoyée nommée 'dataCleaned.csv' se trouvant dans le dossier data 

In [48]:
merged_data.to_csv('data/dataCleaned.csv', index=False, float_format='%.2f')


Analyser les données de vente du CRM et identifier les tendances et les modèles.

identifier les étapes clés où les opportunités sont gagnées ou perdues.

 l'analyse de la distribution des opportunités par étape du pipeline (deal_stage ) pour identifier

les opportunités gagnées ou perdues . en comptant le nombre d'occurrences de chaque valeur unique dans la colonne deal_stage


In [49]:
merged_data["deal_stage"].value_counts()

deal_stage
Won            4238
Lost           2473
Engaging       1589
Prospecting     500
Name: count, dtype: int64

In [50]:
import plotly.express as px

# Compter les occurrences de chaque étape
stage_counts = merged_data["deal_stage"].value_counts().reset_index()
stage_counts.columns = ["deal_stage", "count"]  

#  graphique en entonnoir
fig = px.funnel(
    stage_counts,
    x="count",  # Nombre d'opportunités
    y="deal_stage",  # Étapes du pipeline
    title="Distribution des opportunités par étape  du pipeline de ventes (Deal Stage)",
    labels={"count": "Nombre d'opportunités", "deal_stage": "Étape du pipeline de ventes"}
)

# Afficher le graphique
fig.show()


Analyse vue d'ensemble du graphique

d'après le graphique de notre pipeline de vente  , on remarque que l'étape (Won) à le plus grand nombre d'opportunités(4238) , ce qui signifie que de nombreuses affaires aboutissent avec succès .Avec également une perte considerable de 2473 des opportunités perdues a l'étape (Lost) ,les efforts doivent être faits pour comprendre pourquoi les opportunités sont perdues  et comment améliorer ce taux.
des actions doivent fait pour convertir l'etape (Engaging) qui est de 1589 en opportunités de ventes futurs . avec un goulot d'étranglement dans l'étape(Prospecting) de 500 qui reste faible par rapport aux autres étapes , ce qui signifie que la propection reste un point faible dans le processus 


Analyse stratégique :

Vous pourriez analyser pourquoi certaines opportunités se déplacent d'une étape à l'autre avec succès, tandis que d'autres échouent. Cela peut impliquer une révision des pratiques de prospection ou d'engagement.
Identifier des modèles ou des tendances au niveau des données liées aux opportunités (secteur, type de produit, etc.) peut fournir des insights pour optimiser le pipeline.


Actions et recommandations:

Renforcer la prospection : Mettre en place des outils ou stratégies pour augmenter le nombre d'opportunités dans l'étape "Prospecting".
Réduire les pertes : Analyser les causes de l'étape "Lost" et introduire des mesures pour diminuer les abandons (formation, offres adaptées, etc.).
Améliorer la conversion : Identifier les caractéristiques des opportunités qui progressent avec succès à travers les étapes pour les appliquer aux opportunités en difficulté.

In [51]:
Mt_de_trans_par_deal_stage=merged_data.groupby('deal_stage')['close_value'].sum()

print(Mt_de_trans_par_deal_stage)


deal_stage
Engaging         750008.0
Lost                  0.0
Prospecting      236000.0
Won            10005534.0
Name: close_value, dtype: float64


C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\63408177.py:1: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [52]:
fig = px.pie(
    Mt_de_trans_par_deal_stage.reset_index(),
    names='deal_stage',
    values='close_value',
    title="Répartition des montants des transactions par etape du pipeline de ventes  ",
    hole=0.4, 
    labels={'deal_stage': 'Étape du pipeline', 'close_value': 'Montant total'}
)
fig.show()

Analyse du visuel de la repartition de transaction par etape du pipeline de ventes

Étape majoritaire :

La grande majorité des montants des transactions (91%) se trouvent à l'étape "Won" (gagnée). Cela indique une bonne performance globale dans la conclusion des opportunités de vente.
Étapes moins performantes :

Les montants associés à l'étape "Engaging" représentent 6.82%.
L'étape "Prospecting" est très faible avec seulement 2.15%.
L'étape "Lost" (perdue) est quasiment nulle ou très marginale.

Interprétation :

Une grande proportion de transactions atteignent la phase finale et sont conclues (Won). Cela peut indiquer un pipeline de vente efficace ou des produits/services attractifs.
Cependant, la faible proportion dans les étapes intermédiaires ("Engaging", "Prospecting") peut suggérer :
Un manque de nouvelles opportunités en amont.
Une nécessité d’optimiser les efforts au début du pipeline pour augmenter le nombre de prospects qualifiés.


Recommandations potentielles :

Augmenter les efforts de prospection :
Identifier pourquoi le volume des opportunités est faible au niveau de "Prospecting".
Renforcer les stratégies de marketing et de génération de leads.
Améliorer la conversion dans les étapes intermédiaires :
Identifier les blocages ou obstacles dans l’étape "Engaging".
Former les commerciaux pour optimiser leurs interactions avec les clients potentiels.
Analyser les pertes :
Même si les "Lost" sont faibles, analyser pourquoi certaines opportunités sont perdues pour minimiser ce taux

Analyse de la performances des commerciaux 


In [53]:
revenue_by_agent = merged_data.groupby('sales_agent')['revenue'].sum().sort_values(ascending=False)

print(revenue_by_agent)


sales_agent
Darcel Schlecht       1411515.45
Vicki Laflamme        1266555.52
Kary Hendrixson       1264753.51
Kami Bicknell          956047.87
Zane Levy              895845.96
Anna Snelling          893267.25
Maureen Marcano        841976.20
Markita Hansen         830669.65
Corliss Cosme          808993.81
Jonathan Berthelot     752529.86
Cassey Cress           747226.16
Donn Cantrell          718116.21
James Ascencio         710062.05
Versie Hillebrand      677823.25
Gladys Colclough       648560.96
Lajuana Vencill        648271.29
Daniell Hammack        637191.34
Reed Clapper           625660.61
Violet Mclelland       557859.16
Niesha Huffines        522330.75
Marty Freudenburg      521877.92
Boris Faz              515104.59
Hayden Neloms          505679.40
Moses Frase            481117.35
Elease Gluck           420866.01
Cecily Lampkin         411954.09
Rosie Papadopoulos     408136.23
Rosalina Dieter        362807.61
Garret Kinder          342703.86
Wilburn Farren         232358.1

In [54]:
fig =px.bar(revenue_by_agent, title="revenue_par_commercial",labels={'value':'revenue','sales_agent':'commercial'})

fig.show()

l'analyse des revenue par commercial afin de mettre en lumière ou les efforts peuvent être concentrés pour optimiser la performance globale

Top performeurs :

Les commerciaux Darcel Schlecht, Kary Hendrixson et Vicki Laflamme sont les meilleurs contributeurs en termes de revenus, générant plus de 1 million de revenus chacun.
Cela pourrait indiquer une expertise spécifique, un portefeuille client avantageux ou une stratégie efficace qu'ils emploient.

Performances moyennes :

Un groupe de commerciaux, comme Zane Levy, Kami Bicknell, et Anna Snelling, génèrent des revenus moyens, mais ils ont un potentiel pour atteindre le niveau des top performeurs.

Performances faibles :

Les commerciaux comme Wilburn Farren, Garret Kinder, et Rosalina Dieter génèrent les revenus les plus faibles. Cela pourrait être dû à :
Une zone géographique moins rentable.
Moins d’expérience ou de formation.
Un manque de focus ou de leads de qualité.

Interprétation générale :

La distribution des performances montre une forte concentration de revenus parmi quelques commerciaux, avec un déclin progressif vers les moins performants.
Cela indique une inégalité de performance dans l'équipe commerciale, ce qui peut être un point d'amélioration.

Recommandations suggérés :

Analyser les stratégies des meilleurs performeurs pour les partager avec l'équipe : outils utilisés, méthodes de communication, gestion des leads, etc.

Former les commerciaux avec des performances faibles : Mettre en place des programmes de coaching et des outils pour améliorer leurs résultats.

Répartition des leads : Équilibrer la qualité des leads attribués pour donner des chances égales à tous les commerciaux.

Focus géographique ou sectoriel : Examiner si les secteurs ou régions attribués à certains commerciaux influencent leur performance.



2.3 Analyse de la segmentation par secteur

In [55]:
revenue_by_sector= merged_data.groupby("sector")["revenue"].sum().sort_values(ascending=False)

print(revenue_by_sector)

sector
software              4285554.27
technolgy             4058060.84
Inconnu               2419935.00
retail                2353855.56
medical               1564479.30
finance               1492698.60
telecommunications    1442328.95
marketing             1181885.46
entertainment          876968.17
employment             520220.55
services               421875.38
Name: revenue, dtype: float64


C:\Users\fanti\AppData\Local\Temp\ipykernel_22588\853997489.py:1: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [56]:
fig=px.bar(revenue_by_sector,title='Revenus par secteur', labels={"value":'revenue','sector':'Secteur'})

fig.show()

2.4 Analyse du revenu par secteur 


Observation :
Secteurs dominants :

Les secteurs Software et Technology génèrent les revenus les plus élevés, dépassant les 4 millions d'unités monétaires chacun.
Cela suggère que ces secteurs sont des marchés clés pour l'entreprise et contribuent de manière significative au chiffre d'affaires.
Secteurs intermédiaires :

Les secteurs Retail, Medical, Finance, et Telecommunications génèrent des revenus moyens, situés entre 2 et 3 millions d'unités.
Ces secteurs offrent des opportunités de croissance, mais ne dominent pas encore le pipeline.
Secteurs moins performants :

Les secteurs Marketing, Entertainment, Employment, et Services ont les revenus les plus bas.
Ces secteurs pourraient ne pas représenter des priorités stratégiques ou souffrent d’un manque d’opportunités exploitées.
Catégorie "Inconnu" :

La catégorie "Inconnu" génère un revenu important. Cela peut indiquer des données mal classées ou des clients pour lesquels le secteur n'a pas été renseigné.

Interprétation :

Focus stratégique :
Le pipeline est clairement dominé par les secteurs technologiques. Il serait pertinent de continuer à investir dans ces secteurs tout en explorant des opportunités pour diversifier les sources de revenus.
Opportunités d'amélioration :
Les secteurs à revenus faibles (Marketing, Services, etc.) peuvent être analysés pour comprendre s’ils offrent un potentiel inexploité ou si des efforts doivent être réorientés ailleurs.
Nettoyage des données :
La catégorie "Inconnu" pourrait être examinée de près. Une classification correcte des clients dans les secteurs appropriés peut améliorer les analyses futures.


Recommandations  suggéres:

Renforcer les efforts dans les secteurs dominants :

Continuer à cibler les clients des secteurs Software et Technology en leur offrant des solutions adaptées pour maintenir la position dominante.

Analyser les secteurs moins performants :

Identifier les causes des faibles performances dans les secteurs comme Services et Marketing. Est-ce un problème de marché ou de stratégie de vente ?

Traiter la catégorie "Inconnu" :

Travailler avec les équipes pour compléter les données manquantes et classifier correctement les clients non attribués à un secteur.

Diversifier le pipeline de ventes :

Bien que Software et Technology soient dominants, il est important de diversifier les secteurs cibles pour réduire les risques liés à la dépendance à quelques industries.

In [57]:
merged_data.columns.to_list()

['opportunity_id',
 'sales_agent',
 'product',
 'account',
 'deal_stage',
 'engage_date',
 'close_date',
 'close_value',
 'sector',
 'year_established',
 'revenue',
 'employees',
 'office_location',
 'subsidiary_of',
 'engage_day',
 'engage_month',
 'engage_year',
 'engage_quarter',
 'close_day',
 'close_month',
 'close_year',
 'close_quarter',
 'manager',
 'regional_office',
 'series',
 'sales_price']

In [58]:
print(merged_data.dtypes)



opportunity_id              object
sales_agent                 object
product                     object
account                     object
deal_stage                category
engage_date         datetime64[ns]
close_date          datetime64[ns]
close_value                float64
sector                    category
year_established             int32
revenue                    float64
employees                    int32
office_location             object
subsidiary_of               object
engage_day                  object
engage_month                object
engage_year                  int32
engage_quarter               int32
close_day                   object
close_month                 object
close_year                   int32
close_quarter                int32
manager                     object
regional_office           category
series                      object
sales_price                float64
dtype: object


In [59]:
print(merged_data[['engage_date', 'close_date']].head())

  engage_date close_date
0  2016-10-20 2017-03-01
1  2016-10-25 2017-03-11
2  2016-10-25 2017-03-07
3  2016-10-25 2017-03-09
4  2016-10-25 2017-03-02
